### ISSUE TITLE
CI build problem

### ISSUE THREAD

**Issue by janbartel** [eclipse/jetty.project#8214]
**jetty-12.0.x-eex-jettymavenplugin**

Below is an excerpt from build 9 from PR-8200, on branch jetty-12.0.x-eex-jettymavenplugin:   https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/PR-8200/9/pipeline

There are 2 problems here that _never occur_ on a local build:

1. the WARNING regarding the `ee10-apache-jsp` pom. Nothing `ee10` should even be referenced during a build of the `jetty-ee9-maven-plugin`. No `ee10` artifacts are ever reported by `mvn dependency:tree` nor `mvn -X compile`.
2. the enforcer violation for `org.eclipse.jetty.toolchain:jetty-jakarta-servlet-api` is shown because `ee10` artifacts are being referenced, which does not happen locally (nor should it ever happen!)

> [INFO] ------------< org.eclipse.jetty.ee9:jetty-ee9-maven-plugin >------------
> [INFO] Building EE9 :: Jetty :: Jetty Maven Plugin 12.0.0-SNAPSHOT      [31/53]
> [INFO] ----------------------------[ maven-plugin ]----------------------------
> [WARNING] The POM for org.eclipse.jetty.ee10:jetty-ee10-apache-jsp:jar:12.0.0-20220530.113549-2 is invalid, transitive dependencies (if any) will not be available, enable debug logging for more details
> [INFO] 
> [INFO] --- maven-clean-plugin:3.2.0:clean (default-clean) @ jetty-ee9-maven-plugin ---
> [INFO] 
> [INFO] --- maven-enforcer-plugin:3.0.0:enforce (ban-javax-servlet-api) @ jetty-ee9-maven-plugin ---
> [INFO] 
> [INFO] --- maven-enforcer-plugin:3.0.0:enforce (enforce-java) @ jetty-ee9-maven-plugin ---
> [WARNING] Rule 6: org.apache.maven.plugins.enforcer.RequireUpperBoundDeps failed with message:
> Failed while enforcing RequireUpperBoundDeps. The error(s) are [
> Require upper bound dependencies error for org.eclipse.jetty.toolchain:jetty-jakarta-servlet-api:5.0.2 paths to dependency are:
> 
> +-org.eclipse.jetty.ee9:jetty-ee9-maven-plugin:12.0.0-SNAPSHOT
>   +-org.eclipse.jetty.ee9:jetty-ee9-annotations:12.0.0-SNAPSHOT
>     +-org.eclipse.jetty.toolchain:jetty-jakarta-servlet-api:5.0.2 (managed) <-- org.eclipse.jetty.toolchain:jetty-jakarta-servlet-api:5.0.2
> 
> and
> 
> +-org.eclipse.jetty.ee9:jetty-ee9-maven-plugin:12.0.0-SNAPSHOT
>   +-org.eclipse.jetty.ee9:jetty-ee9-apache-jsp:12.0.0-SNAPSHOT
>     +-org.eclipse.jetty.toolchain:jetty-jakarta-servlet-api:5.0.2 (managed) <-- org.eclipse.jetty.toolchain:jetty-jakarta-servlet-api:5.0.2
> 
> and
> 
> +-org.eclipse.jetty.ee9:jetty-ee9-maven-plugin:12.0.0-SNAPSHOT
>   +-org.eclipse.jetty.ee9:jetty-ee9-servlet:12.0.0-SNAPSHOT
>     +-org.eclipse.jetty.ee9:jetty-ee9-nested:12.0.0-SNAPSHOT (managed) <-- org.eclipse.jetty.ee9:jetty-ee9-nested:12.0.0-SNAPSHOT
>       +-org.eclipse.jetty.toolchain:jetty-jakarta-servlet-api:5.0.2 (managed) <-- org.eclipse.jetty.toolchain:jetty-jakarta-servlet-api:5.0.2
> 
> and
> 
> +-org.eclipse.jetty.ee9:jetty-ee9-maven-plugin:12.0.0-SNAPSHOT
>   +-org.eclipse.jetty:jetty-home:12.0.0-SNAPSHOT [test]
>     +-org.eclipse.jetty.ee10:jetty-ee10-servlet:12.0.0-SNAPSHOT [test]
>       +-org.eclipse.jetty.toolchain:jetty-jakarta-servlet-api:5.0.2 [test] (managed) <-- org.eclipse.jetty.toolchain:jetty-jakarta-servlet-api:6.0.0-SNAPSHOT [test]
> 
> and
> 
> +-org.eclipse.jetty.ee9:jetty-ee9-maven-plugin:12.0.0-SNAPSHOT
>   +-org.eclipse.jetty:jetty-home:12.0.0-SNAPSHOT [test]
>     +-org.eclipse.jetty.ee10.websocket:jetty-ee10-websocket-jetty-server:12.0.0-SNAPSHOT [test]
>       +-org.eclipse.jetty.toolchain:jetty-jakarta-servlet-api:5.0.2 [test] (managed) <-- org.eclipse.jetty.toolchain:jetty-jakarta-servlet-api:6.0.0-SNAPSHOT [test]
> 
> and
> 
> +-org.eclipse.jetty.ee9:jetty-ee9-maven-plugin:12.0.0-SNAPSHOT
>   +-org.eclipse.jetty:jetty-home:12.0.0-SNAPSHOT [test]
>     +-org.eclipse.jetty.ee10:jetty-ee10-proxy:12.0.0-SNAPSHOT [test]
>       +-org.eclipse.jetty.toolchain:jetty-jakarta-servlet-api:5.0.2 [test] (managed) <-- org.eclipse.jetty.toolchain:jetty-jakarta-servlet-api:6.0.0-SNAPSHOT [test]
> 
> and
> 
> +-org.eclipse.jetty.ee9:jetty-ee9-maven-plugin:12.0.0-SNAPSHOT
>   +-org.eclipse.jetty:jetty-home:12.0.0-SNAPSHOT [test]
>     +-org.eclipse.jetty.ee10:jetty-ee10-annotations:12.0.0-SNAPSHOT [test]
>       +-org.eclipse.jetty.toolchain:jetty-jakarta-servlet-api:5.0.2 [test] (managed) <-- org.eclipse.jetty.toolchain:jetty-jakarta-servlet-api:6.0.0-SNAPSHOT [test]

**Comment by joakime** [comment]
BTW, this is why I want to keep the `ee#` trees isolated from each other.
Right now in `jetty-12.0.x` things are messy/muddy.

**Comment by joakime** [comment]
Looks to be coming from the use of `jetty-home`, which hasn't been cleaned up to use the overlay concepts I proposed.
(The overlay concepts would break the transitive deep dives into the `ee#` trees from the use of `jetty-home`)

**Comment by joakime** [comment]
I think the individual `jetty-ee#-maven-plugin` implementations should have no `jetty-home` support.
That's nonsense to me.

A top level `jetty-home-maven-plugin` should exist for that.

**Comment by joakime** [comment]
The build on CI is still using the "lets limp along" configuration.

https://github.com/eclipse/jetty.project/blob/jetty-12.0.x-eex-jettymavenplugin/Jenkinsfile#L21

This builds in the following order ...

``` shell
mvn clean install -f build
mvn -N clean install 
mvn clean install -f jetty-core
mvn clean install -f jetty-integrations
mvn clean install -f jetty-ee9
mvn clean install -f jetty-ee8
mvn clean install -f jetty-ee10
# mvn clean install -f jetty-home
# mvn clean install -f tests
# mvn clean install -f documentation
```

If you take this Dockerfile and do the above order you'll be able to replicate.

``` Dockerfile
FROM eclipse-temurin:17-jdk-focal

# Default to UTF-8 file.encoding
ENV LANG C.UTF-8

# Packages Update / Install

RUN apt-get -q update && \
    apt-get -q install -y \
               git && \
    apt-get -q autoremove && \
    apt-get -q clean y && \
    rm -rf /var/lib/apt/lists/* && \
    rm -rf /var/cache/apt/*.bin

# Run as Builder User

ENV HOME /home/builder
RUN groupadd -g 10000 builder
RUN useradd -c "Builder user" -d $HOME -u 10000 -g 10000 -m builder

USER builder
WORKDIR $HOME
RUN mkdir $HOME/code

# Install maven

ENV MAVEN_VERSION=3.8.6

RUN curl --no-progress-meter \
         --location \
         --output $HOME/apache-maven-$MAVEN_VERSION-bin.tar.gz \
         https://dlcdn.apache.org/maven/maven-3/$MAVEN_VERSION/binaries/apache-maven-$MAVEN_VERSION-bin.tar.gz && \
    tar -zxf $HOME/apache-maven-$MAVEN_VERSION-bin.tar.gz && \
    rm -f $HOME/apache-maven-$MAVEN_VERSION-bin.tar.gz && \
    echo "export M2_HOME=\$HOME/apache-maven-$MAVEN_VERSION" >> $HOME/.bashrc && \
    echo "export PATH=\$M2_HOME/bin:\$PATH" >> $HOME/.bashrc

# Clone jetty-12.0.x branch

WORKDIR $HOME/code

RUN git clone --depth 1 --branch jetty-12.0.x https://github.com/eclipse/jetty.project.git $HOME/code/jetty.project
```

To run ...

```
$ docker build --tag jetty:builder .
```

To run ...

```
$ docker run --rm -it jetty:builder bash
builder@64b0e822ed97:~/code$ cd jetty.project/
builder@64b0e822ed97:~/code/jetty.project$ mvn clean install -f build
[INFO] Scanning for projects...
... (etc) ...
```

**Comment by joakime** [comment]
To get past CI we can go back to the simpler CI build.  (See PR #8217 for simpler build against this branch)

Ultimately, we should not be depending on `jetty-home` in any fashion from within the various `jetty-ee#/` directories.

**Comment by janbartel** [comment]
OK. So it seems that  when running locally, all projects are either in the reactor together, or have already previously been built. However, CI builds projects separately. Therefore, we have to avoid the situation where `eex-maven-plugin` depends on `jetty-home`, because `jetty-home` cannot be built until `jetty-eex` is built, and `jetty-eex` cannot be built until `jetty-home` is built.

I have a solution/solutions for that:

1. `eex-maven-plugin` does not need to depend on `jetty-home` for any reason other than sequencing the build, so the direct dependency on that can be removed.
2. all of the integration tests of the `eex-maven-plugin`s must be moved to the `tests` module, which is built _after_ `jetty-home`, so the artifact will be available for the aether RepositorySystem to dynamically resolve at runtime. 

I've checked these changes into the `jetty-12.0.x-eex-jettymavenplugin` branch, lets see if that makes CI happy.

**Comment by janbartel** [comment]
I've merged #8200, with the changes as outlined above.  CI was green on the PR build, however I suspect that the `tests` project is not really being executed. @olamy or @joakime can one of you with sufficient jenkins foo ensure that `tests` is built to `install` stage, including integration test phase?

**Comment by janbartel** [comment]
CI has been updated and is now building all modules correctly.
